### PR TITLE
kt: Retry request if remote end performs graceful restart

### PR DESCRIPTION
**Pull request by lmb**
A HTTP1 server closes all idle connections when it performs a graceful shutdown. This is a problem for things like http.Transport, which maintain a pool of (possibly idle) connections, since an attempt to re-use such a closed connection will return io.EOF to the caller. This makes it impossible to implement zero-downtime restarts which do not drop requests.

This change treats io.EOF as a signal that a zero-downtime restart is in progress. It closes all other idle connections, and retries the request which yielded io.EOF once.

### CODE REVIEW

File: kt/kt.go [hunk 1]
```diff
@@ -315,6 +302,41 @@ func (c *Conn) doRPC(path string, values []KV) (code int, vals []KV, err error)
...
+	}
+	return resp, t, nil
+}
+
+func (c *Conn) makeRequest(method string, url *url.URL, headers http.Header, body []byte) (*http.Request, *time.Timer) {
+	req := &http.Request{
+		Method:        method,
+		URL:           url,
+		Header:        headers,
+		Body:          ioutil.NopCloser(bytes.NewBuffer(body)),
```
Comment (FiloSottile) [inline kt/kt.go]: This should be a `bytes.NewReader`. (I know it was wrong before, but while you are at it...)

File: kt/kt.go [hunk 2]
```diff
@@ -315,6 +302,41 @@ func (c *Conn) doRPC(path string, values []KV) (code int, vals []KV, err error)
...
+		ContentLength: int64(len(body)),
+	}
+	t := time.AfterFunc(c.timeout, func() {
+		c.transport.CancelRequest(req)
```
Comment (FiloSottile) [inline kt/kt.go]: This doesn't work with H/2, you could replace it with a much cleaner `Request.WithContext(context.WithTimeout)`

Comment (FiloSottile) [inline kt/kt.go]: Also, I noticed you set `ResponseHeaderTimeout`. If that timeout hits and this doesn't, an error is returned which is not `ErrTimeout`. I suggest you switch to just the context timeout, and remove `ResponseHeaderTimeout`.

Comment (lmb) [inline kt/kt.go]: Good point. I'll defer to Pavel whether he wants that in this change or not.

File: kt/kt.go (hunk 1 above)
Comment (lmb) [inline kt/kt.go]: Fixed

**Comment by pavel-odintsov** [comment]
LGTM!

**Pull request merged**